### ISSUE TITLE
[dart2js] Crash trying to report diagnostics

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/sdk#52020]
dart2js used to report late errors (e.g. from ssa/builder.dart), but now crashes.
This makes working in the runtime libraries harder.

The internal libraries contain calls to many special functions that can only be checked in the back-end.

Repro: Change `lib/_internal/js_runtime/lib/js_helper.dart`, at line 2139
```diff
-      final typeEvalMethod = RAW_DART_FUNCTION_REF(BoundClosure.evalRecipe);
+      final typeEvalMethod = RAW_DART_FUNCTION_REF(()=>functionType);
```
Then `build.py -mrelease`

Old error:
```
Error: 'RAW_DART_FUNCTION_REF' requires a static method or top-level method.
      final typeEvalMethod = RAW_DART_FUNCTION_REF(()=>functionType);
                             ^
```

New error:
```
Uncaught exception in diagnostic handler: Unsupported operation: Cannot extract a file path from a org-dartlang-sdk URI
#0      _SimpleUri.toFilePath (dart:core/uri.dart:4843:7)
#1      SourceFileProvider._readFromFileSync (package:compiler/src/source_file_provider.dart:72:48)
#2      SourceFileProvider.getUtf8SourceFile (package:compiler/src/source_file_provider.dart:117:11)
#3      FormattingDiagnosticHandler.report (package:compiler/src/source_file_provider.dart:257:45)
#4      Compiler.callUserHandler.<anonymous closure> (package:compiler/src/compiler.dart:895:17)
#5      CompilerTask.measure (package:compiler/src/common/tasks.dart:66:51)
#6      Compiler.callUserHandler (package:compiler/src/compiler.dart:894:23)
#7      Compiler._reportDiagnosticMessage (package:compiler/src/compiler.dart:886:7)
#8      Compiler.reportDiagnostic (package:compiler/src/compiler.dart:873:5)
#9      DiagnosticReporter._reportDiagnostic (package:compiler/src/diagnostics/diagnostic_listener.dart:122:15)
#10     DiagnosticReporter._reportDiagnosticInternal (package:compiler/src/diagnostics/diagnostic_listener.dart:117:5)
#11     DiagnosticReporter.reportError (package:compiler/src/diagnostics/diagnostic_listener.dart:46:5)
#12     DiagnosticReporter.reportErrorMessage (package:compiler/src/diagnostics/diagnostic_listener.dart:51:5)
#13     KernelSsaGraphBuilder._handleForeignRawFunctionRef (package:compiler/src/ssa/builder.dart:4807:14)
#14     KernelSsaGraphBuilder._handleInvokeStaticForeign (package:compiler/src/ssa/builder.dart:4479:7)
...
```

I think this started happening with 35492bdf141c096e9bb790b84ae8e6ce497d2af1

/cc @natebiggs

**Comment by biggs0125** [comment]
I see I didn't think we'd be emitting errors like this on internal libraries. As the error says it doesn't know how to read the `org-dartlang-sdk` scheme in the URI for internal libraries.

Since this is only an issue for internal libraries I think we can get away with leaving off the actual source code. Errors could look like this:
```
org-dartlang-sdk:///lib/_internal/js_runtime/lib/js_helper.dart@77853+1:
Error: 'RAW_DART_FUNCTION_REF' requires a static method or top-level method.
```
It doesn't contain the exact source but it gives you the file offset which is easy enough to look up.

Otherwise we would *always* have to cache all non-`file` scheme URIs which seems wasteful for the 99.99...% of builds that have no chance of encountering these errors.

Proposed fix here: https://dart-review.googlesource.com/c/sdk/+/294720

**Comment by sigmundch** [comment]
Oh, unfortunately using offset is also a regression for something we fixed recently.

We often get crash reports from the wild, and it's not rare for the top failure to originate from an SDK library location (examples: [1][1], [2][2]). Offset based errors w/o context are harder to diagnose. It requires pinning the exact SDK version, check out the associated SDK sources, to then map to the correct line location. If however the location comes form the flutter SDK, this gets a lot more complicated.

Our fix at the time was to add contents to the cache, which is exactly what you are trying to remove. Take a look at https://dart-review.googlesource.com/c/sdk/+/251700

I can imagine a couple options here:
* we restore the cache, but under a flag, so we would keep it clear in our internal bazel builds where it provides no value, but use the cache externally?
* find an alternative approach to help us diagnose things faster?

I must say, it was nice when we landed that CL to be able to finally say that we were no longer going to have offset-based error messages, and that all messages had a clear line/column and code snippet from the error location. That makes me lean towards the flag approach.

[1]: https://github.com/dart-lang/sdk/issues/33898
[2]: https://github.com/dart-lang/sdk/issues/41491
